resources/type-system/inference.md

```diff
@@ -0,0 +1,830 @@
+  - The greatest closure of `S` with respect to `L` is `Future<U>` where `U` is
+    the greatest closure of `T` with respect to `L`
+- if `S` is `FutureOr<T>`
+  - The least closure of `S` with respect to `L` is `FutureOr<U>` where `U` is the
+    least closure of `T` with respect to `L`
+  - The greatest closure of `S` with respect to `L` is `FutureOr<U>` where `U` is
+    the greatest closure of `T` with respect to `L`
+- if `S` is an interface type `C<T0, ..., Tk>`
+  - The least closure of `S` with respect to `L` is `C<U0, ..., Uk>` where `Ui`
+    is the least closure of `Ti` with respect to `L`
+  - The greatest closure of `S` with respect to `L` is `C<U0, ..., Uk>` where
+    `Ui` is the greatest closure of `Ti` with respect to `L`
+- if `S` is `U Function<X0 extends B0, ...., Xk extends Bk>(T0 x0, ...., Tn xn,
+  [Tn+1 xn+1, ..., Tm xm])` and `L` contains any free type variables from any of
+  the `Bi`:
+  - The least closure of `S` with respect to `L` is `Null`
+  - The greatest closure of `S` with respect to `L` is `Object`
+- if `S` is `T Function<X0 extends B0, ...., Xk extends Bk>(T0 x0, ...., Tn xn,
+  [Tn+1 xn+1, ..., Tm xm])` and `L` does not contain any free type variables
+  from any of the `Bi`:
+  - The least closure of `S` with respect to `L` is `U Function<X0 extends B0,
+  ...., Xk extends Bk>(U0 x0, ...., Un1 xn, [Un+1 xn+1, ..., Um xm])` where:
+    - `U` is the least closure of `T` with respect to `L` 
+    - `Ui` is the greatest closure of `Ti` with respect to `L` 
+    - with the usual capture avoiding requirement that the `Xi` do not appear in
+  `L`.
+  - The greatest closure of `S` with respect to `L` is `U Function<X0 extends B0,
+  ...., Xk extends Bk>(U0 x0, ...., Un1 xn, [Un+1 xn+1, ..., Um xm])` where:
+    - `U` is the greatest closure of `T` with respect to `L` 
+    - `Ui` is the least closure of `Ti` with respect to `L` 
+    - with the usual capture avoiding requirement that the `Xi` do not appear in
+  `L`.
+- if `S` is `T Function<X0 extends B0, ...., Xk extends Bk>(T0 x0, ...., Tn xn,
+  {Tn+1 xn+1, ..., Tm xm})` and `L` does not contain any free type variables
+  from any of the `Bi`:
+  - The least closure of `S` with respect to `L` is `U Function<X0 extends B0,
+  ...., Xk extends Bk>(U0 x0, ...., Un1 xn, {Un+1 xn+1, ..., Um xm})` where:
+    - `U` is the least closure of `T` with respect to `L` 
+    - `Ui` is the greatest closure of `Ti` with respect to `L` 
+    - with the usual capture avoiding requirement that the `Xi` do not appear in
+  `L`.
+  - The greatest closure of `S` with respect to `L` is `U Function<X0 extends B0,
+  ...., Xk extends Bk>(U0 x0, ...., Un1 xn, {Un+1 xn+1, ..., Um xm})` where:
+    - `U` is the greatest closure of `T` with respect to `L` 
+    - `Ui` is the least closure of `Ti` with respect to `L` 
+    - with the usual capture avoiding requirement that the `Xi` do not appear in
+  `L`.
+
+
+# MATERIAL BELOW HERE HAS NOT BEEN UPDATED #
```

Comment:
I have a bunch of comments on stuff below, but I'll postpone that as well (noting that you might have plans for a more extensive revision of this material, in which case my comments could soon become obsolete).

Maybe it would be useful to land the material up to this point, and then just indicate here that there will be more material added later on? With that, we could say that we have a well-defined chunk of the inference specification, and then we can work on the rest (and maybe add more chunks in whatever order fits best).